api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java

```diff
@@ -479,4 +485,384 @@ private DateObsCohortDefinition getLastArvPickupDateCohort() {
+    cd.setOperator(SetComparator.IN);
+    return cd;
+  }
+
+  /**
+   * Combine the viral load and viral load qualitative
+   *
+   * @return CohortDefinition
+   */
+  private CohortDefinition getViralLoadOrQualitative() {
+    CompositionCohortDefinition cd = new CompositionCohortDefinition();
+    cd.setName("Combined viral load and its qualitative patients");
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+    cd.addSearch(
+        "VL",
+        map(
+            getViralLoadTestDone(),
+            "onOrAfter=${startDate},onOrBefore=${endDate},locationList=${location}"));
+    cd.addSearch(
+        "VLQ",
+        map(
+            gePatientsWithCodedObs(hivMetadata.getHivViralLoadQualitative()),
+            "onOrAfter=${startDate},onOrBefore=${endDate},locationList=${location}"));
+    cd.setCompositionString("VL OR VLQ");
+    return cd;
+  }
+
+  /**
+   * E2 Get the combinations together E2: Number of active patients in ART at the end of current
+   * month who received a Viral Load Test Result (Annual Notification
+   *
+   * @return CohortDefinition
+   */
+  public CohortDefinition
+      getNumberOfActivePatientsInArtAtTheEndOfTheCurrentMonthHavingVlTestResults() {
+    CompositionCohortDefinition cd = new CompositionCohortDefinition();
+    cd.setName(
+        "E2: Number of active patients in ART at the end of current month who received a Viral Load Test Result (Annual Notification");
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+    cd.addSearch(
+        "C",
+        map(
+            getStandardDefinitionForEcolumns(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "VL",
+        map(
+            getViralLoadOrQualitative(),
+            "startDate=${startDate},endDate=${endDate},location=${location}"));
+    cd.addSearch(
+        "VLX",
+        map(
+            getPatientsToExclude(
+                hivMetadata.getHivViralLoadConcept(),
+                hivMetadata.getAdultoSeguimentoEncounterType(),
+                "NUMERIC",
+                "NO",
+                "obs"),
+            "onOrAfter=${startDate},onOrBefore=${endDate},location=${location}"));
+    cd.addSearch(
+        "VLQ",
+        map(
+            getPatientsToExclude(
+                hivMetadata.getHivViralLoadQualitative(),
+                hivMetadata.getAdultoSeguimentoEncounterType(),
+                "CODED",
+                "NO",
+                "obs"),
+            "onOrAfter=${startDate},onOrBefore=${endDate},location=${location}"));
+    cd.setCompositionString("(C AND VL) AND (VLX OR VLQ)");
```

Comment:
This composition seems wrong: should it be: `(C AND VL) AND NOT (VLX OR VLQ)`?

note the exclusion...